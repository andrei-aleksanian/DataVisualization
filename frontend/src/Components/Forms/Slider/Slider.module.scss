.index {
  margin: 0 0 3rem 0;
}

.index p {
  font-size: var(--font-size-label-p);
  margin: 0 0 var(--margin-under-labels) 0;
}
