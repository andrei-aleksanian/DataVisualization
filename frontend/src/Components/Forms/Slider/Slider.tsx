/*
Slider Component.

All intended uses imply marks on the slider, but the slider is highly customizable.
See rc-slider https://www.npmjs.com/package/rc-slider for more details.

Unfortunately, this library doesn't support css modules very well,
so all css is hardcoded inside style elements.
*/
import { CSSProperties } from 'react';

import SliderRC from 'rc-slider';
import classes from './Slider.module.scss';
import toMarks from './utils';

export interface SliderProps {
  onChange: (value: number) => void;
  min: number;
  max: number;
  step: number | null;
  marksArr: (number | string)[];
  text: string;
  value: number;
}

const Slider = ({ marksArr, text, ...restProps }: SliderProps) => {
  const marksStyle = {
    fontSize: '0.85rem',
    color: '#ccc',
  } as CSSProperties;

  const handleStyle = {
    borderColor: 'var(--color-action)',
    height: 20,
    width: 20,
    marginTop: -9,
    backgroundColor: 'var(--color-action)',
    boxShadow: 'none',
  };
  const railStyle = { backgroundColor: 'var(--color-text)', height: 4 };
  const trackStyle = { backgroundColor: 'var(--color-action)', height: 4 };
  const dotStyle = { border: '2px solid #fff', height: 10, width: 10, bottom: -3 };
  const activeDotStyle = { border: '2px solid var(--color-action)', backgroundColor: 'var(--color-action)' };

  return (
    <div className={classes.index}>
      <p>{text}</p>
      <SliderRC
        {...restProps}
        marks={toMarks(marksArr, marksStyle)}
        handleStyle={handleStyle}
        railStyle={railStyle}
        trackStyle={trackStyle}
        dotStyle={dotStyle}
        activeDotStyle={activeDotStyle}
      />
    </div>
  );
};

export default Slider;
