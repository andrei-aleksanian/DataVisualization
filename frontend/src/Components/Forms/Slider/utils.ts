import { CSSProperties, ReactNode } from 'react';

export interface Marks {
  [number: number]: { style: CSSProperties; label: ReactNode };
}

const toMarks = (data: (string | number)[], style: CSSProperties) => {
  /*
  Converts data like ['1','10%'] to:
  {0:{style: CSSProperties, label: 1}, 1:{style: CSSProperties, label: '10%'}} (0 and 1 are indexes of teh input array)
  and data like [1,2] to:
  {1:{style: CSSProperties, label: 1}, 2:{style: CSSProperties, label: 2}}

  Used in the Slider Component internally to allow ease of input in the parent components.
  */

  const marks: Marks = {};
  for (let i = 0; i < data.length; i += 1) {
    const currentMark = data[i];
    if (typeof currentMark === 'string') {
      marks[i] = {
        style,
        label: data[i],
      };
    } else {
      marks[currentMark] = {
        style,
        label: data[i],
      };
    }
  }

  return marks;
};

export default toMarks;
