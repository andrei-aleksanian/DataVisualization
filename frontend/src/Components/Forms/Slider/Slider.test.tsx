import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Slider, { SliderProps } from './Slider';
import toMarks from './utils';

let state = 0;
const setState = (value: number) => {
  state = value;
};

const getSliderProps = (num: boolean = true): SliderProps => {
  const testSliderProps: SliderProps = {
    onChange: setState,
    min: 0,
    max: 0.3,
    step: 0.1,
    marksArr: [],
    text: 'Test Description',
    value: state,
  };
  if (num) {
    testSliderProps.marksArr = [0, 0.1, 0.2, 0.3];
  } else {
    testSliderProps.marksArr = ['0', '0.1', '0.2', '0.3'];
  }
  return testSliderProps;
};
const cleanupState = () => {
  state = 0;
};

const init = () => {
  render(<Slider {...getSliderProps()} />);
};

describe('Test Slider Component with marks number inputs', () => {
  test('Matches snapshot', async () => {
    const { asFragment } = render(<Slider {...getSliderProps()} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('Description is rendered', () => {
    init();
    expect(screen.getByText(getSliderProps().text)).toBeInTheDocument();
  });

  test('Change state on click', () => {
    init();
    userEvent.click(screen.getByText('0.1'));
    expect(state).toBe(0.1);
    userEvent.click(screen.getByText('0.2'));
    expect(state).toBe(0.2);

    cleanupState();
  });
});

describe('Test Slider Utils', () => {
  const style = {
    color: 'red',
  };

  test('toMarks converts number array correctly', () => {
    const marksArr = [0, 1, 2, 3];
    const marksObj = toMarks(marksArr, style);
    for (let i = 0; i < marksArr.length; i += 1) {
      const mark = marksArr[i];
      expect(marksObj[mark].label).toBe(mark);
      expect(marksObj[mark].style).toBe(style);
    }
  });

  test('toMarks converts string array correctly', () => {
    const marksArr = ['0', '1', '2', '3'];
    const marksObj = toMarks(marksArr, style);
    for (let i = 0; i < marksArr.length; i += 1) {
      const mark = marksArr[i];
      expect(marksObj[i].label).toBe(mark);
      expect(marksObj[i].style).toBe(style);
    }
  });
});
