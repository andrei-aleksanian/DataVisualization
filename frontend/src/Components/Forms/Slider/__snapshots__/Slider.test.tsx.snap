// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Slider Component with marks number inputs Matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="index"
  >
    <p>
      Test Description
    </p>
    <div
      class="rc-slider rc-slider-with-marks"
    >
      <div
        class="rc-slider-rail"
        style="height: 4px;"
      />
      <div
        class="rc-slider-track"
        style="height: 4px; left: 0%; width: 0%;"
      />
      <div
        class="rc-slider-step"
      >
        <span
          class="rc-slider-dot rc-slider-dot-active"
          style="height: 10px; width: 10px; bottom: -3px; left: 0%;"
        />
        <span
          class="rc-slider-dot"
          style="border: 2px solid #fff; height: 10px; width: 10px; bottom: -3px; left: 33.333333333333336%;"
        />
        <span
          class="rc-slider-dot"
          style="border: 2px solid #fff; height: 10px; width: 10px; bottom: -3px; left: 66.66666666666667%;"
        />
        <span
          class="rc-slider-dot"
          style="border: 2px solid #fff; height: 10px; width: 10px; bottom: -3px; left: 100%;"
        />
      </div>
      <div
        aria-disabled="false"
        aria-valuemax="0.3"
        aria-valuemin="0"
        aria-valuenow="0"
        class="rc-slider-handle"
        role="slider"
        style="height: 20px; width: 20px; margin-top: -9px; box-shadow: none; left: 0%; transform: translateX(-50%);"
        tabindex="0"
      />
      <div
        class="rc-slider-mark"
      >
        <span
          class="rc-slider-mark-text rc-slider-mark-text-active"
          style="transform: translateX(-50%); left: 0%; font-size: 0.85rem; color: rgb(204, 204, 204);"
        >
          0
        </span>
        <span
          class="rc-slider-mark-text"
          style="transform: translateX(-50%); left: 33.333333333333336%; font-size: 0.85rem; color: rgb(204, 204, 204);"
        >
          0.1
        </span>
        <span
          class="rc-slider-mark-text"
          style="transform: translateX(-50%); left: 66.66666666666667%; font-size: 0.85rem; color: rgb(204, 204, 204);"
        >
          0.2
        </span>
        <span
          class="rc-slider-mark-text"
          style="transform: translateX(-50%); left: 100%; font-size: 0.85rem; color: rgb(204, 204, 204);"
        >
          0.3
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
